oci/private/pull.bzl

```diff
@@ -199,28 +203,87 @@ def _oci_pull_impl(rctx):
+                rctx.attr.registry,
+                rctx.attr.repository,
+            ))
 
         matching_manifest = _find_platform_manifest(manifest, rctx.attr.platform)
         if not matching_manifest:
-            fail("No matching manifest found in image {}/{} for platform {}".format(rctx.attr.registry, rctx.attr.repository, rctx.attr.platform))
-
-        # extra download to get the manifest for the target platform
+            fail("No matching manifest found in image {}/{} for platform {}".format(
+                rctx.attr.registry,
+                rctx.attr.repository,
+                rctx.attr.platform,
+            ))
         manifest, size, digest = downloader.download_manifest(matching_manifest["digest"], "manifest.json")
     else:
         fail("Unrecognized mediaType {} in manifest file".format(manifest["mediaType"]))
-    
-    # symlink manifest.json to blobs with it's digest. 
+
+    # symlink manifest.json to blobs with its digest.
     # it is okay to use symlink here as copy_to_directory will dereference it when creating the TreeArtifact.
     rctx.symlink("manifest.json", _digest_into_blob_path(digest))
 
-    # download the image config
-    downloader.download_blob(manifest["config"]["digest"], _digest_into_blob_path(manifest["config"]["digest"]))
+    config_output_path = _digest_into_blob_path(manifest["config"]["digest"])
+    downloader.download_blob(manifest["config"]["digest"], config_output_path)
+
+    # if the user provided a platform for the image, validate it matches the config as best effort.
+    if rctx.attr.platform:
+        config_bytes = rctx.read(config_output_path)
+        config = json.decode(config_bytes)
+
+        image_os = config.get("os", None)
+        image_architecture = config.get("architecture", None)
+        image_variant = config.get("variant", None)
+
+        attr_os, _, attr_architecture = rctx.attr.platform.partition("/")
+        attr_variant = None
+        if "/" in attr_architecture:
+            attr_architecture, _, attr_variant = attr_architecture.partition("/")
+
+        # the OCI spec makes os and architecture fields required but the Docker one doesn't specify.
+        missing_fields = []
+        if not image_os:
+            missing_fields.append("os")
+        elif image_os != attr_os:
+            fail("Expected image {}/{} to have os '{}', got: '{}'".format(
+                rctx.attr.registry,
+                rctx.attr.repository,
+                attr_os,
+                image_os,
+            ))
+        if not image_architecture:
+            missing_fields.append("architecture")
+        elif image_architecture != attr_architecture:
+            fail("Expected image {}/{} to have architecture '{}', got: '{}'".format(
+                rctx.attr.registry,
+                rctx.attr.repository,
+                attr_architecture,
+                image_architecture,
+            ))
```

Comment:
I believe this would be a nice utility function imho. i'd like to keep the implementation function as lean as possible.  Goal is https://github.com/bazel-contrib/rules_oci/issues/138

Reply:
Good point, done 👍